### PR TITLE
[AD-135] Finish Ariadne.Wallet.Backend.Restore

### PR DESCRIPTION
**Pull request by Pastafarianist**
TODO:
- [x] Allow custom account names
- [x] Check that the root exists in the DB in `createHdAccount`

### CODE REVIEW

File: ariadne/src/Ariadne/Wallet/Cardano/Kernel/DB/AcidState.hs [hunk 1]
```diff
@@ -151,6 +152,7 @@ applyBlock (blocksByAccount,meta) = runUpdateNoErrors $ zoom dbHdWallets $
         (\prefBlock -> zoom hdAddressCheckpoints $
                            modify $ Spec.applyBlock (prefBlock,meta))
         blocksByAccount
+        Map.empty -- we are not expecting to create new accounts in applyBlock
```
Comment (gromakovsky) [inline ariadne/src/Ariadne/Wallet/Cardano/Kernel/DB/AcidState.hs]: What does it mean?
```
-- | For each of the specified accounts, create them if they do not exist,
-- and apply the specified function.
createPrefiltered
```

AFAIU, the whole purpose of `createPrefiltered` is to create new accounts, isn't it?

File: ariadne/src/Ariadne/Wallet/Backend/AddressDiscovery.hs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+-- | This is heavily based on @Pos.Crypto.HDDiscovery.discoverHDAddresses@.
+discoverHDAddressesWithUtxo
+    :: (MonadDBRead m, MonadUnliftIO m)
+    => NonEmpty HDPassphrase
+    -> m (NonEmpty AddressWithPathToUtxoMap)
+discoverHDAddressesWithUtxo walletPassphrases =
+    runConduitRes $ utxoSource .| CL.fold step initWallets
+  where
+    initWallets = map (const Map.empty) walletPassphrases
```
[the remark this answers is not in the document]
Comment (Pastafarianist) [reply ariadne/src/Ariadne/Wallet/Backend/AddressDiscovery.hs]: I think I'm good.

File: ariadne/src/Ariadne/Wallet/Cardano/Kernel/DB/AcidState.hs (hunk 1 above)
Comment (Pastafarianist) [reply ariadne/src/Ariadne/Wallet/Cardano/Kernel/DB/AcidState.hs]: `createPrefiltered` is a complicated helper. It basically takes a `Map HdAccountId PrefilteredUtxo` and some auxiliary instructions on what to do with the accounts & addresses after ensuring that they are created. I guess the way it was brought into existence is that someone at IOHK noticed that there were two very similar chunks of code in `applyBlock` and `createHdWallet`, so they extracted that code into a separate function. As long as this function is not exposed outside of the module, it remains an implementation detail, which is OK IMO.

File: ariadne/src/Ariadne/Wallet/Backend/AddressDiscovery.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (neongreen) [reply ariadne/src/Ariadne/Wallet/Backend/AddressDiscovery.hs]: doesn't everyone

[the remark this answers is not in the document]
Comment (Pastafarianist) [reply ariadne/src/Ariadne/Wallet/Backend/AddressDiscovery.hs]: It's not August 8 yet, not even July 8.

[the remark this answers is not in the document]
Comment (neongreen) [reply ariadne/src/Ariadne/Wallet/Backend/AddressDiscovery.hs]: :unamused:

File: ariadne/src/Ariadne/Wallet/Cardano/Kernel/DB/AcidState.hs (hunk 1 above)
Comment (Pastafarianist) [reply ariadne/src/Ariadne/Wallet/Cardano/Kernel/DB/AcidState.hs]: > the `createPrefiltered` function does create new accounts (at least sometimes)

This is true.

> But you say that we are not expecting to create new accounts. If we are not expecting it, why do we call `createPrefiltered`?

Like I said, `createPrefiltered` is a complicated helper. It both ensures that all the necessary accounts and addresses exist, and applies updates to them. In `applyBlock`, the _updates_ part is important. In `createHdWallet`, _creation_ matters.

> After ensuring that they have even been created, not in this particular call?

Yes.

File: ariadne/src/Ariadne/Wallet/Backend/AddressDiscovery.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Pastafarianist) [reply ariadne/src/Ariadne/Wallet/Backend/AddressDiscovery.hs]: Come back in a week

**Pull request merged**